File: frameset.go
package fileseq

import (
	"fmt"
	"math"

	"github.com/justinfx/gofileseq/v3/ranges"
)

const (
	// Maximum frames for general ranges (simple, x modifier, : stagger modifier)
	maxRangeSize = 100000000 // 100 million

	// Y modifier has O(n²) performance (AppendUnique with Contains checks)
	// so it needs a much lower limit
	maxYModifierRange = 1000000 // 1 million

	// Maximum iterations for : (stagger) modifier
	maxStaggerIterations = 10000

	// Maximum chunk value for y modifier (prevents overflow in skip calculation)
	maxChunkValue = 1000000000 // 1 billion
)

// checkRangeSize validates that a range size doesn't exceed the given maximum
func checkRangeSize(start, end, step int, maxSize int64, context string) error {
	if step == 0 {
		step = 1
	}

	diff := math.Abs(float64(end - start))
	size := int64(math.Ceil(diff / math.Abs(float64(step))))

	if size > maxSize {
		return fmt.Errorf("%s: range %d-%d (step %d) would create %d frames, max allowed is %d",
			context, start, end, step, size, maxSize)
	}
	return nil
}

// FrameSet wraps a sequence of frames in container that
// exposes array-like operations and queries, after parsing
// the given frame range string.
type FrameSet struct {
	frange   string
	rangePtr *ranges.InclusiveRanges
}

// NewFrameSet creates a new FrameSet from a given frame range string
// Returns an error if the frame range could not be parsed.
func NewFrameSet(frange string) (*FrameSet, error) {
	// Process the frame range and get a slice of match slices
	matches, err := frameRangeMatches(frange)
	if err != nil {
		return nil, err
	}

	frameSet := &FrameSet{frange, &ranges.InclusiveRanges{}}

	// Process each slice match and add it to the frame set
	for _, match := range matches {
		if err = frameSet.handleMatch(match); err != nil {
			return nil, err
		}
	}

	return frameSet, nil
}

// Process a rangePattern match group
func (s *FrameSet) handleMatch(match []string) error {
	switch len(match) {

	// Single frame match
	case 1:
		f, err := parseInt(match[0])
		if err != nil {
			return err
		}
		s.rangePtr.AppendUnique(f, f, 1)

	// Simple frame range
	case 2:
		start, err := parseInt(match[0])
		if err != nil {
			return err
		}
		end, err := parseInt(match[1])
		if err != nil {
			return err
		}

		// Handle descending frame ranges, like 10-1
		var inc int
		if start > end {
			inc = -1
		} else {
			inc = 1
		}

		// Check range size before appending
		if err := checkRangeSize(start, end, inc, maxRangeSize, "frame range too large"); err != nil {
			return err
		}

		s.rangePtr.AppendUnique(start, end, inc)

	// Complex frame range
	case 4:
		var (
			err               error
			mod               string
			start, end, chunk int
		)
		chunk, err = parseInt(match[3])
		if err != nil {
			return err
		}
		if chunk == 0 {
			return fmt.Errorf("Failed to parse part of range %v. "+
				"Encountered invalid 0 value", match[3])
		}
		if start, err = parseInt(match[0]); err != nil {
			return err
		}
		if end, err = parseInt(match[1]); err != nil {
			return err
		}
		if mod = match[2]; !isModifier(mod) {
			return fmt.Errorf("%q is not one of the valid modifier 'xy:'", mod)
		}

		switch mod {
		case `x`:
			// Check range size before appending
			if err := checkRangeSize(start, end, chunk, maxRangeSize, "frame range too large"); err != nil {
				return err
			}
			s.rangePtr.AppendUnique(start, end, chunk)

		case `y`:
			// Y modifier has O(n²) performance due to AppendUnique Contains checks
			if err := checkRangeSize(start, end, 1, maxYModifierRange, "y modifier range too large"); err != nil {
				return err
			}

			// Limit chunk value to prevent integer overflow in skip calculation
			if chunk > maxChunkValue || chunk < -maxChunkValue {
				return fmt.Errorf("y modifier value too large: %d, max allowed is %d", chunk, maxChunkValue)
			}

			// TODO: Add proper support for adding inverse of range.
			// This approach will add excessive amounts of single
			// range elements. They could be compressed into chunks
			skip := start
			aRange := ranges.NewInclusiveRange(start, end, 1)
			var val int
			for it := aRange.IterValues(); !it.IsDone(); {
				val = it.Next()
				if val == skip {
					skip += chunk
					continue
				}
				s.rangePtr.AppendUnique(val, val, 1)
			}

		case `:`:
			// Check range size for stagger modifier
			if err := checkRangeSize(start, end, 1, maxRangeSize, "stagger range too large"); err != nil {
				return err
			}

			// Limit stagger iterations to prevent excessive processing
			// Each iteration appends the full range with a different step
			if chunk > maxStaggerIterations {
				return fmt.Errorf("stagger modifier value too large: %d, max allowed is %d",
					chunk, maxStaggerIterations)
			}

			for ; chunk > 0; chunk-- {
				s.rangePtr.AppendUnique(start, end, chunk)
			}
		}

	default:
		return fmt.Errorf("Unexpected match []string size: %v", match)
	}

	return nil
}

// String implements Stringer, by returning the frame
// range string
func (s *FrameSet) String() string {
	frange := s.FrameRange()
	return frange
}

// Len returns the number of frames in the set
func (s *FrameSet) Len() int {
	return s.rangePtr.Len()
}

// Index returns the index position of the frame value
// within the frame set.
// If the given frame does not exist, then return -1
func (s *FrameSet) Index(frame int) int {
	return s.rangePtr.Index(frame)
}

// Frame returns the frame number value for a given index into
// the frame set.
// If the index is outside the bounds of the frame set range,
// then an error is returned.
func (s *FrameSet) Frame(index int) (int, error) {
	return s.rangePtr.Value(index)
}

// Frames returns a slice of the frame numbers that were parsed
// from the original frame range string.
//
// Warning: This allocates a slice containing number of elements
// equal to the Len() of the range. If the sequence is massive,
// it could hit slice limits. Better to use IterFrames()
func (s *FrameSet) Frames() []int {
	size := s.rangePtr.Len()
	frames := make([]int, size, size)
	i := 0
	for it := s.rangePtr.IterValues(); !it.IsDone(); {
		frames[i] = it.Next()
		i++
	}
	return frames
}

// HasFrame returns true if the frameset contains the given
// frame value.
func (s *FrameSet) HasFrame(frame int) bool {
	return s.rangePtr.Contains(frame)
}

// Start returns the first frame in the frame set
func (s *FrameSet) Start() int {
	return s.rangePtr.Start()
}

// End returns the last frame in the frame set
func (s *FrameSet) End() int {
	return s.rangePtr.End()
}

// FrameRange returns the range string that was used
// to initialize the FrameSet
func (s *FrameSet) FrameRange() string {
	return s.frange
}

// FrameRangePadded returns the range string that was used
// to initialize the FrameSet, with each number padded out
// with zeros to a given width
func (s *FrameSet) FrameRangePadded(pad int) string {
	return PadFrameRange(s.frange, pad)
}

// Invert returns a new FrameSet that represents
// all frames *not* within the current FrameSet. That is, it
// will create a range that "fills in" the current one.
func (s *FrameSet) Invert() *FrameSet {
	ptr := s.rangePtr.Inverted()
	return &FrameSet{ptr.String(), ptr}
}

// InvertedFrameRange returns a new frame range string that represents
// all frames *not* within the current FrameSet. That is, it
// will create a range that "fills in" the current one.
func (s *FrameSet) InvertedFrameRange(pad int) string {
	frange := s.rangePtr.Inverted().String()
	if pad > 1 {
		frange = PadFrameRange(frange, pad)
	}
	return frange
}

// Normalize returns a new sorted and compacted FrameSet
func (s *FrameSet) Normalize() *FrameSet {
	ptr := s.rangePtr.Normalized()
	return &FrameSet{ptr.String(), ptr}
}

File: cpp/frameset.cpp
#include "frameset.h"
#include "error.h"
#include "pad.h"

#include "private/frameset_p.h"

#include <algorithm>
#include <iostream>
#include <sstream>

namespace fileseq {


FrameSet::FrameSet() : m_frameData(new internal::FrameSetData) {

}


FrameSet::FrameSet(const std::string &frange, Status* ok)
    : m_frameData(new internal::FrameSetData(frange))
{
    Status stat;
    internal::RangeMatches matches;

    // Process the frame range and get a list of matches
    if (!(stat = internal::frameRangeMatches(matches, frange))) {
        if (ok != nullptr) {
            *ok = stat;
        }
        setInvalid();
        return;
    }

    // Process each list of matches and add it to the frame set
    for (auto &matche : matches) {

        // If handleMatch succeeds, it will populate the FrameSet
        // with more parts of the range
        if (handleMatch(&matche, &stat), !stat) {

            // We could not process a match for some reason
            if (ok != nullptr) {
                *ok = stat;
            }
            setInvalid();
            return;
        }
    }

    // A valid frameset
    if (ok != nullptr) ok->clearError();

    // Save the original frame range string
    m_frameData->frameRange = frange;
}


FrameSet::~FrameSet() {
    setInvalid();
}


FrameSet::FrameSet(const FrameSet& other)
    : m_frameData(new internal::FrameSetData) {

    if (other.m_frameData != nullptr) {
        (*m_frameData) = *(other.m_frameData);
    }
}


void FrameSet::setInvalid() {
    if ( m_frameData != nullptr) {
        delete m_frameData;
        m_frameData = nullptr;
    }
}


void FrameSet::handleMatch(const internal::RangePatternMatch* match, Status* ok) {
    const int num = match->matches;

    ok->clearError();

    Frame start, end;
    long step = 1;

    switch (num) {

    // Single frame match
    case 1:
        start = match->start;
        m_frameData->ranges.appendUnique(start, start, 1);
        return;

    // Simple frame range
    case 2:
        start = match->start;
        end = match->end;

        // Handle descending frame ranges, like 10-1
        step = (start > end) ? -1 : 1;

        m_frameData->ranges.appendUnique(start, end, step);
        return;

    // Complex frame range
    case 4:
        start = match->start;
        end = match->end;
        step = match->step;

        if (step == 0) {
            ok->setError("Got invalid step value 0");
            return;
        }

        if (!internal::isRangeModifier(match->stepMod)) {
            std::stringstream ss;
            ss << match->stepMod << " is not one of the valid modifiers 'xy:'";
            ok->setError(ss.str());
        }

        char mod = match->stepMod[0];

        switch (mod) {

        case 'x':
            m_frameData->ranges.appendUnique(start, end, step);
            break;

        case 'y': {
            // TODO: Add proper support for adding inverse of range.
            // This approach will add excessive amounts of singe
            // range elements. They could be compressed into chunks
            Frame skip = start;
            Range aRange(start, end, 1);
            RangeIterator it = aRange.iterValues();
            Frame val;
            while (it.next()) {
                val = (*it);
                if (val == skip) {
                    skip += step;
                    continue;
                }
                m_frameData->ranges.appendUnique(val, val, 1);
            }
            break;
        }

        case ':':
            while (step > 0) {
                m_frameData->ranges.appendUnique(start, end, step);
                step--;
            }
            break;

        }

        return;
    }

    // If we get here, we  have not handled the match
    std::stringstream ss;
    ss << "Unexpected match size: " << num;
    ok->setError(ss.str());
}

bool FrameSet::isValid() const {
    if ( m_frameData == nullptr) {
        return false;
    }

    return m_frameData->ranges.length() != 0;

}

std::string FrameSet::string() const {
    return isValid() ? m_frameData->frameRange : "";
}

size_t FrameSet::length() const {
    return isValid() ? m_frameData->ranges.length() : 0;
};

size_t FrameSet::index(Frame frame) const {
    return isValid() ? (size_t)m_frameData->ranges.index(frame) : 0;
};

Frame FrameSet::frame(size_t index, Status* ok) const {
    if (!isValid()) {
        return 0;
    }

    return m_frameData->ranges.value(index, ok);
};

void FrameSet::frames(Frames &frames) const {
    frames.clear();

    if (!isValid()) {
        return;
    }

    size_t len = length();
    if (len <= 0) {
        return;
    }

    frames.reserve(len);

    RangesIterator it = m_frameData->ranges.iterValues();
    while (it.next()) {
        frames.push_back(*it);
    }
};

RangesIterator FrameSet::iterFrames() const {
    if (!isValid()) {
        return RangesIterator();
    }

    return m_frameData->ranges.iterValues();
}

bool FrameSet::hasFrame(Frame frame) const {
    return isValid() ? m_frameData->ranges.contains(frame) : false;
};

Frame FrameSet::start() const {
    return isValid() ? m_frameData->ranges.start() : 0;
};

Frame FrameSet::end() const {
    return isValid() ? m_frameData->ranges.end() : 0;
};

std::string FrameSet::frameRange(int pad) const {
    if (!isValid()) {
        return "";
    }

    if (pad < 2) {
        return m_frameData->frameRange;
    }

    return padFrameRange(m_frameData->frameRange, (size_t)pad);
};

FrameSet FrameSet::inverted() const {
    FrameSet newFrameSet;

    if (!isValid()) {
        return newFrameSet;
    }

    // get a new range that is inverted
    Ranges ranges;
    m_frameData->ranges.inverted(ranges);

    // Create a new internal data member and swap in our ranges
    newFrameSet.m_frameData->frameRange = ranges.string();
    std::swap(newFrameSet.m_frameData->ranges, ranges);

    return newFrameSet;
};

std::string FrameSet::invertedFrameRange(int pad) const {
    if (!isValid()) {
        return "";
    }

    Ranges ranges;
    m_frameData->ranges.inverted(ranges);

    std::string frange = ranges.string();

    if (pad > 1) {
        frange = padFrameRange(frange, (size_t)pad);
    }

    return frange;
};

FrameSet FrameSet::normalized() const {
    FrameSet newFrameSet;

    if (!isValid()) {
        return newFrameSet;
    }

    // get a new range that is normalized
    Ranges ranges;
    m_frameData->ranges.normalized(ranges);

    // Create a new internal data member and swap in our ranges
    newFrameSet.m_frameData->frameRange = ranges.string();
    std::swap(newFrameSet.m_frameData->ranges, ranges);

    return newFrameSet;
};

} // fileseq
